### PR TITLE
Drawing: Add tasks to drawing tools

### PR DESCRIPTION
**Pull request by eatyourgreens**
Add tool.tasks to the Tool model. 
- Allow question and text tasks as drawing tool tasks. 
- Add `tool.hasCompletedTasks` to check for complete annotations on subtasks.
- Copy task tests from the Step model.

Package:
lib-classifier

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
I’m pretty sure we’ve already discussed this in an RFC for the new classifier. I’ll look it up tomorrow.

**Comment by eatyourgreens** [comment]
The reason being that Caesar has trouble with tool details.

**Comment by eatyourgreens** [comment]
There’s no changes to the lab since this is only an internal representation of the tool tasks within the classifier’s models. The payload received from Panoptes won’t need to change.

**Comment by srallen** [comment]
I don't see this proposed change as clearly discussed in any of the possibly related RFC issues: 
- Transcription sub-task: https://github.com/zooniverse/front-end-monorepo/issues/1314#issuecomment-561717018
- Drawing sub-task: https://github.com/zooniverse/front-end-monorepo/issues/1239
- Mark models: https://github.com/zooniverse/front-end-monorepo/issues/500

I've asked for clarification from @CKrawczyk but there are a few other items to discuss such as which kinds of tasks the sub-task should support as well. I really think we need to be sure we discuss this in issue first, then write an ADR if it's a change affecting architecture, which I believe this is, then write the code. It's difficult to ask for documentation after code is written and I think we should get better at practicing doing documentation first as a team.

**Comment by eatyourgreens** [comment]
I’ll look it up tomorrow and see what we decided. It was basically that subtasks annotations should be represented in exactly the same way as classification annotations ie. subtasks should have task keys. They don’t in PFE.

**Comment by eatyourgreens** [comment]
This isn’t a replacement for `details`. For each task snapshot in `details`, we will add a task key, create a Task and then put the task in the `tasks` array. It’s a purely internal representation of tool tasks using the existing classifier models.

**Comment by srallen** [comment]
Ok, thanks for some clarification. I'm finding it hard to follow what is for what, which is why I keep asking for documentation prior to work beginning. Is there an issue that this is linked to?

**Comment by eatyourgreens** [comment]
Just the issues about adding tasks and their annotations to drawing marks.

**Comment by eatyourgreens** [comment]
Here you go. https://github.com/zooniverse/front-end-monorepo/issues/1239#issuecomment-561732576

**Comment by eatyourgreens** [comment]
Reading back through the code, the logic for validating a tool here is completely wrong, which is my fault for copying it directly from the Step model. The spec for validating a tool with subtasks should be:
- a mark is complete when it has annotations for each required tool task.
- a tool is complete when all of its marks are complete.

**Comment by eatyourgreens** [comment]
@srallen Here's where we discussed changing tool details: https://github.com/zooniverse/front-end-monorepo/issues/123#issuecomment-418108424

**Comment by srallen** [comment]
A single comment in an old closed RFC issue focused on a different issue and solution, even if there is some relation, isn't a discussion. A decision on the sub-task wasn't made in ADR 5. 

I'd prefer we finalize the plan for sub-tasks in its relevant RFC, https://github.com/zooniverse/front-end-monorepo/issues/1239, and then commit the decision as an ADR, so it is easy to reference, follow, and remember for the whole team.

**Comment by eatyourgreens** [comment]
Sounds good to me. I've been updating #1239 with comments about progress, and planned next steps, anyway.

This PR is just focussed on adding tasks to the state tree so that we can render them in the subject viewer. I do think there is a separate discussion that needs to be had fairly soon about how the annotations from those tasks will be serialised and posted to Panoptes/Caesar/any other service that wants to use them. #500 seems like the natural place to have that discussion. https://github.com/zooniverse/front-end-monorepo/issues/500#issuecomment-564448229

By the way, I'm fine with renaming `tool.tasks` to `tool.details` but I think it will be a little confusing since its type will be `types.array(Task)`. We'll probably also end up with code like:
`tool.details.map(task => <TaskComponent task={task} />` which also might not be clear to a first-time viewer who doesn't know that tasks are called details.

**Comment by srallen** [comment]
I'm fine with a renaming when there's a good reason and we document that in an ADR, but it's serialized from Panoptes as `details` so how would applying a response as a snapshot work with this renaming? We can have a breaking change, but right now I don't think this will work without a mapping from one to another and then later a plan to update the lab to change `details` to `tasks`. If we're going to do a breaking change, we could also plan on flattening the task structure on the workflow side as well as on the annotation side similar to what's discussed in #500.

**Comment by eatyourgreens** [comment]
`details` is invalid as a Snapshot because the Task model requires an identifier and details don’t have task keys. Same reason workflow tasks can’t be applied directly as a snapshot.

**Comment by srallen** [comment]
I'm confused by your response. Workflow tasks do have unique identifiers, T0, T1, etc, so my question still remains. How does this work with a response from Panoptes?

Doing a quick check with a project that has a drawing tool with a sub-task created, here's what the response looks like from Panoptes with a sub-task defined in details:

Screen Shot 2019-12-11 at 1 15 42 PM

And how it's stored in the stores:

Screen Shot 2019-12-11 at 1 15 13 PM

tasks is empty.

Either we need to match the data structure coming from Panoptes or have a plan on how this is changing with documentation and likely some kind of helper function to make the conversion like we do with tasks to steps as a temporary interim step until the lab can be updated.

**Comment by eatyourgreens** [comment]
Sorry, I meant  workflow tasks don't have a `taskKey` property, which is required by the Task model. There's a line in one of the classifier models that adds this property to new tasks before creating them.

**Comment by eatyourgreens** [comment]
Looking at the workflow model, to see how workflow tasks were modelled, I think `tool.details` will be `types.array[types.frozen]` and then `tool.tasks` will be an array of Task models built from the details array.

I’ll read up, though, to see if there’s a way of applying a snapshot via a function that turns it into a valid Task snapshot.

**Comment by eatyourgreens** [comment]
Oh, and tasks is supposed to be empty in this PR. It doesn’t add any code to create the task models.

**Comment by srallen** [comment]
Before diving too much into that, I think the option of flattening the task structure should be considered. I'm in favor of doing this on the workflow model side like we're considering with the annotations.

**Comment by eatyourgreens** [comment]
From a mark, I’d like to access subtasks as `mark.tool.tasks` but that could be a computed view, I suppose. See my earlier comment about calculating completeness for marks and their parent tools. 

My reservation about storing tasks elsewhere in the state tree would be that it introduces a dependency that makes tools and marks hard to mock and test. We’d have to mock a workflow as well, I think.

**Comment by eatyourgreens** [comment]
References may help here.

**Comment by eatyourgreens** [comment]
#1372 sets up tools as ancestors of marks in the tree, which kind of points where I’m going with this.

**Comment by eatyourgreens** [comment]
Closing this in favour of a PR where I spec the behaviour first.
